Image Upload 기능 구현

olivejua opened the pull request:
관련이슈: #18 

Post 첨부이미지와 User 프로필 이미지 업로드 기능을 구현한다.

olivejua committed:
:see_no_evil: gitignore: application-aws.yml 추가 (#18)

추가 수정사항: `application-aws.yml`이 `main`과 `test` 경로 둘 다 존재하기 때문에 `*.yml` 파일 모두 상세경로 덧붙여줌

olivejua committed:
:wrench: config: application.yml에 profile.include 값 추가 (#18)

olivejua committed:
:heavy_plus_sign: dependency: spring-cloud-starter-aws 의존성 추가 (#18)

S3 업로드 기능 구현 목적

olivejua committed:
:sparkles: feature: S3 스토리지에 파일 업로드 및 삭제 기능 구현 (#18)

AmazonS3Config.java: S3 클라우드에 접근할 수 있는 Client 객체를 생성
S3Uploader.java: S3에 파일을 업로드 및 삭제하는 로직

olivejua committed:
:sparkles: feature: S3 스토리지에 2개 이상의 객체 삭제 기능 구현 및 주석 추가 (#18)

olivejua committed:
:recycle: refactor: S3Uploader IOException 처리 로직 추가 (#18)

메서드를 호출하는 곳마다 IOException을 throw하는 것을 방지하는 목적

olivejua committed:
:sparkles: feature: 엔티티 Post, User에 Image 정보를 담고있는 Embedded 객체 추가 (#18)

- Post: PostImages JoinTable 추가
- User: Profile (profileMessage, profileImage를 담고 있는) 객체 생성

olivejua committed:
:sparkles: feature: 각 도메인별 상황에 맞춘 UploadService 객체 생성 (#18)

- ImageUtils 생성: Image 형식 검증 로직과 Image 경로 정보를 담고 있다.
- UploadService 생성: 이 인터페이스를 통해 이미지 업로드 및 삭제 기능을 사용함으로 확장성 있게 만든다.
- S3UploadService 생성: 구체적인 업로드 구현로직을 담고 있고 S3에 의존하고 있다.

olivejua committed:
:white_check_mark: test: Image 검증 로직과 업로드 기능을 테스트한다 (#18)

- ImageUtilsTest: 이미지 검증 로직 테스트
- UploadServiceTest: Post, User 이미지 업로드 기능을 테스트한다.
- testImage.jpg 추가: 실제 업로드기능을 테스트하기 위한 테스트 이미지

viiviii commented on .gitignore:
```diff
@@ -4,10 +4,14 @@ HELP.md
-application-db.yml
+src/main/resources/application-oauth.yml
+src/main/resources/application-db.yml
+src/main/resources/application-aws.yml
```
ignore 할 파일이 여러 경로로 존재할 때 상세 경로 붙인 이유가 경로를 더 명시적으로 나타내기 위함인가용?

olivejua replied:
@viiviii 네 main과 test 환경 모두 적용한다는 점이 좀 더 직관적으로 와닿는것 같아서 수정했어요!

viiviii commented on src/main/java/com/simplelog/api/commons/upload/S3UploadService.java:
```diff
@@ -0,0 +1,89 @@
+
+    public int removeAll(Post post) {
+        PostImages images = post.getImages();
+        if (!images.isNotEmpty()) {
```
`!images.isNotEmpty()`보단 `images.isEmpty()`로 하는 게 좋을 것 같아요!(작업하다 깜빡한 것 같아서 적음)

olivejua replied:
예리하네여... 풀리퀘해놓고 코드 다시보다가 `isEmpty` 로 할걸 했어 ㅋㅋㅋㅋㅋㅋ 수정하겠슴니다

viiviii commented on src/main/java/com/simplelog/api/commons/upload/S3UploadService.java:
```diff
@@ -0,0 +1,89 @@
+                .collect(Collectors.toList());
+    }
+
+    private String extractKey(String url, String domainPath) {
```
`ImageUtils`의 `makeImagePath()`가 S3 key를 조합하는 메서드던데 `S3UploadService`의 `extractKey()`와 한 곳에 세트로 있는건 어떨까요?

viiviii commented on src/main/java/com/simplelog/api/commons/upload/S3Uploader.java:
```diff
@@ -0,0 +1,105 @@
+@Component
+public class S3Uploader {
+
+    private final AmazonS3Client amazonS3Client;
```
`AmazonS3Client` 대신 interface인 `AmazonS3`로 선언하는 건 어떨까요?

https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3.html

olivejua replied:
구현클래스보다 인터페이스에 접근하는게 좋겠네요! 수정했습니다.

viiviii commented on src/main/java/com/simplelog/api/commons/upload/S3Uploader.java:
```diff
@@ -0,0 +1,105 @@
+        log.info("파일이 삭제되지 못했습니다.");
+    }
+
+    private Optional<File> convert(MultipartFile file) {
+        try {
+            File convertFile = new File(System.getProperty("user.dir") + "/" + file.getOriginalFilename());
+            if (convertFile.createNewFile()) {
+                try (FileOutputStream fos = new FileOutputStream(convertFile)) {
+                    fos.write(file.getBytes());
+                }
+                return Optional.of(convertFile);
+            }
+        } catch (IOException e) {
+            log.warn("error: MultipartFile -> File convert fail");
+        }
+
+        return Optional.empty();
+    }
```
나중에 `FileConverter` 같은 클래스로 추출하면 어떨까요? (작명 솜씨 미안🍎)

olivejua replied:
지금은 `S3Uploader`에서만 이 기능을 사용하고 길이도 길지 않으니 클래스 분리까지는 아직 이르다고 생각했어! File 처리 관련된 책임이 좀 더 늘어나게 되면 분리도 고려해보겠습니다! 😊

viiviii commented on src/main/java/com/simplelog/api/commons/upload/S3Uploader.java:
```diff
@@ -0,0 +1,105 @@
+        return uploadImageUrl;
+    }
+
+    private String putS3(File uploadFile, String fileName) {
+        amazonS3Client.putObject(new PutObjectRequest(bucket, fileName, uploadFile).withCannedAcl(CannedAccessControlList.PublicRead));
+        return amazonS3Client.getUrl(bucket, fileName).toString();
+    }
```
`String fileName`은 `S3UploadService`에서 생성한 `getImagePath()`에 대한 값이고 다른 메서드에선 `savedPath`라는 변수명을 사용하던데 S3Upload 관련 클래스 내에서는 용어를 하나로 통일하는 건 어떨까요?
저는 `extractKey()`라는 메서드도 있고 S3의 `PutObjectRequest` 클래스에도 `fileName`이 key 값으로 들어가는 거니 `key`로 하는 것도 괜찮을 것 같아요!

olivejua replied:
미처 생각하지 못한 부분이네요! `key`로 통일하는 것 괜찮은 의견인것 같아요! 반영하겠습니다.

olivejua replied:
더불어 `S3UploadService`는 `S3Uploader`를 의존하고 있지만 직접적으로 S3에 의존하는 것은 아니니 S3 용어보다는 일반적인 용어가 더 가독성이 좋을 것 같아 `extractKey()` 에서 `extractPathFromUrl()` 로 메서드명을 변경하겠습니다!

viiviii commented on src/main/java/com/simplelog/api/commons/upload/S3UploadService.java:
```diff
@@ -0,0 +1,89 @@
+     * Remove
+     **************************************************************************************************/
+
+    public void remove(User user) {
+        Profile userProfile = user.getProfile();
+        if (!userProfile.existImage()) {
+            log.info("삭제할 이미지가 없습니다.");
+            return;
+        }
+
+        String key = extractKey(userProfile.getImageUrl(), USER_PROFILE_PATH);
+        s3Uploader.remove(key);
+        user.removeProfileImage();
+    }
```
`remove()`의 매개변수로 `User` 객체가 전달되지만 실제 사용되는 곳을 보면 `Profile` 객체만 사용하고 있는데 그럼 이 메서드는 `Profile`과 상호작용하는 게 아닐까용? 
```
profile.existImage();
profile.getImageUrl();
user.removeProfileImage() // 내부적으로 profile.removeImageUrl() 호출 
```
다른 메서드였다면 `Profile`을 전달 받는 게 좋을 것 같지만, 지금 `User - Profile`이 영한님 강의 중에 `Order-OrderItem`처럼 사용되는 거라면 `user.hasProfileImage()` , `user.getProfileImageUrl()` 처럼 사용하는 건 어떨까요?👻

++ 밑에 Post도!

olivejua replied:
좋은생각인것 같아요!  `S3UploadService`가 `Post, PostImage`와 `User, Profile` 을 모두 아는 것보다 `Post`와 `User`만 알면 간접적으로 접근하니까 좀더 객체지향적일 것 같네요 👍 이점도 반영하겠습니다.

viiviii commented on src/main/java/com/simplelog/api/config/AmazonS3Config.java:
```diff
@@ -0,0 +1,31 @@
+    @Bean
+    public AmazonS3Client amazonS3Client() {
+        BasicAWSCredentials awsCreds = new BasicAWSCredentials(accessKey, secretKey);
+        return (AmazonS3Client) AmazonS3ClientBuilder.standard()
```
여기도 `AmazonS3Client `대신 `AmazonS3` interface로 선언하면 어떨까용😷

viiviii commented on src/main/java/com/simplelog/api/domain/Post.java:
```diff
@@ -62,4 +65,8 @@ public Long getId() {
     public PostTags getPostTags() {
         return postTags;
     }
+
+    public PostImages getImages() {
+        return images;
+    }
```
`PostTages`와 `PostImages` 클래스 명이 세트인 것 처럼 메서드 및 변수명도 `getTags - getImages`나 `getPostTags - getPostImages`로 맞추는 것도 좋아보여요!

olivejua replied:
헉 그러네요 이것도 수정할게요!!

viiviii commented on src/main/java/com/simplelog/api/domain/PostImages.java:
```diff
@@ -0,0 +1,35 @@
+            joinColumns = @JoinColumn(name = "post_id"))
+    @BatchSize(size = 30)
+    @Column(name = "image_url", nullable = false)
+    private List<String> imageUrls;
+
+    public PostImages() {
+        imageUrls = new ArrayList<>();
+    }
```
컬렉션은 필드 레벨에서 바로 초기화 하는건 어떨까요?

```java
private List<String> imageUrls = new ArrayList<>();
```
저도 종종 고민했는데 킹영한님의 강의에서 컬렉션은 필드에서 바로 초기화하는게 안전하고 간결하다고 하셨던 기억이 납니다!
* [해당 블로그의 3번 내용](https://responsibility.tistory.com/56)

olivejua replied:
수정했습니다!

viiviii commented on src/main/java/com/simplelog/api/domain/Profile.java:
```diff
@@ -0,0 +1,41 @@
+        this.imageUrl = imageUrl;
+    }
+
+    public void updateImageUrl(String url) {
+        imageUrl = url;
+    }
+
+    public void removeImageUrl() {
+        imageUrl = null;
+    }
```
값 타입은 immutable 하게 설계하는게 좋대요:-)
영한님 강의에서 embedded인 Address 기준으로 변경을 원하면 새로 생성해서 바꾸라고 설명 하셨던 것!

```java
Address address = new Address("서울", oldAdress.getStreet(), ...);
member.setAddress(address);
```

olivejua replied:
이 부분 수정할게요!

viiviii commented on src/main/java/com/simplelog/api/domain/User.java:
```diff
@@ -26,25 +22,40 @@
+    private Profile profile;
 
     public User(String nickname, String email, String socialCode, String profileMessage) {
         this.nickname = nickname;
         this.email = email;
         this.socialCode = socialCode;
-        this.profileMessage = profileMessage;
+        this.profile = new Profile(profileMessage);
     }
```
생성자에서 new 연산자로 `Collaborator`를 생성하는건 테스트 하기 어려움, SRP 위반 등 의 여지가 있다고 들었어요!👂

* [해당 내용 더 자세한 고수님의 링크 첨부합니다!](https://brunch.co.kr/@myner/63)

olivejua replied:
링크확인했습니다! 링크에서의 예제는 객체 그대로 받지만 저희 같은 경우는 `Post` 객체를 참조해야해서 아직 생성되지 않은 `Post`를 참조하고 객체를 `Post`의 생성자 파라미터로 받는 건 불가능하다고 판단하여 생성자 파라미터에는 `PostTags`와 `PostImages`를 빼고 add 메서드에 파라미터를 변경했습니다.

viiviii commented on src/main/java/com/simplelog/api/utils/ImageUtils.java:
```diff
@@ -0,0 +1,50 @@
+
+    public static String getImagePath(Post post, MultipartFile file) {
+        return makeImagePath(POST_IMAGE_PATH, post.getId(), file.getOriginalFilename());
+    }
```
메서드 시그니처에는 실제로 사용되는 의존성을 나타내는게 좋대! `getImagePath()` 메서드보단 `makeImagePath()`처럼👍
`getImagePath()`를 쓰던 곳에서 `makeImagePath()`를 사용하면 어떨까?

이유 및 자세한 설명은 내가 도움받았던 블로그 글 링크로 대신할게!(난 digging collaborators나 기만적인 API로 외우고있음)
* https://brunch.co.kr/@myner/64

viiviii commented on src/main/java/com/simplelog/api/utils/ImageUtils.java:
```diff
@@ -0,0 +1,50 @@
+
+import java.util.List;
+import java.util.UUID;
+
```
ImageUtils 클래스 명을 좀 더 좁은 의미의 클래스명으로 바꾸면 어떨까요?
넓은 의미를 사용할 경우 🚨**미래의 저**🚨같은 사람이 야금야금 기능을 추가해서 God-Class가 됩니다ㅋㅋㅋ
....음...ImagePathUtils나 S3KeyMaker 같은...? (작명 센스 미안해오...🍎 당장 생각이 안나)

viiviii commented on src/main/java/com/simplelog/api/commons/upload/UploadService.java:
```diff
@@ -0,0 +1,38 @@
+import org.springframework.web.multipart.MultipartFile;
+
+import java.util.List;
+
```
`UploadService` interface가 `User`, `Post` 엔티티를 알고 있어서 확장이 어렵진 않을까요?

`UploadService - S3UploadService`가 예전 MVC 배울 때 Service interface - ServiceImpl class 관계라면, 
`Uploader` interface를 추가해서 `UploaderService`에 DI 하는건 어떨까요? (`S3Uploader`는 `Uploader` Impl)

제가 이해한건 MVC로 비교하면 `S3Uploader`가 AWS s3와 직접 통신해서 crud 해주는 persistance layer 느낌이고 
`UploadService`는 persistance layer를 사용해서 로직을 수행하는 느낌이 아닐까해서 이렇게 상상해봤읍니다 💤

viiviii commented on src/main/java/com/simplelog/api/commons/upload/S3UploadService.java:
```diff
@@ -0,0 +1,89 @@
+     * Upload
+     **************************************************************************************************/
+
+    public String upload(User user, MultipartFile uploadFile) {
+        String savedPath = getImagePath(user, uploadFile);
+        String uploadedUrl = s3Uploader.upload(uploadFile, savedPath);
+
+        user.updateProfileImage(uploadedUrl);
+        return uploadedUrl;
+    }
+
+    public List<String> upload(Post post, List<MultipartFile> uploadFiles) {
+        List<String> uploadedUrls = new ArrayList<>();
+
+        for (MultipartFile uploadFile : uploadFiles) {
+            String savedPath = getImagePath(post, uploadFile);
+            String imageUrl = s3Uploader.upload(uploadFile, savedPath);
+            uploadedUrls.add(imageUrl);
+        }
+
+        post.addImages(uploadedUrls);
+        return uploadedUrls;
+    }
```
이미지 업로드 시 User와 Post 엔티티 상태 변경도 S3UploadService에서 담당하는 이유가 어떤건지 궁금해용!☺️
처음 봤을 때 userService에서 uploadService를 사용할 것 같았는데...
제가 이미지 업로드를 한번도 생각해보거나 공부한 적이 없어서 학습 차원으로 알고싶습니다🥲

viiviii commented on src/main/java/com/simplelog/api/utils/ImageUtils.java:
```diff
@@ -0,0 +1,50 @@
+public abstract class ImageUtils {
+    private final static String SEP = "/";
+
+    public final static String USER_PROFILE_PATH = "user" + SEP + "profile" + SEP;
+    public final static String POST_IMAGE_PATH = "post" + SEP;
```
나중에 이 path들도 property로 관리 해볼까?

olivejua replied:
좋아욥 ㅎㅎ `DomainImagePaths.java` 의 책임이 여기서 더 추가되거나 `Path`가 추가된다면 `Path`들만 따로 빼서 관리하는 방법도 고려해보아요!

viiviii reviewed (commented):
슬기띠 이미지 업로드 구현 수고했어요🥳
리뷰 시 테스트도 함께 본다던데 제가 아직 테스트에 무지하여 드릴 말씀이 없읍니다.. 미안해오😭😭😭

olivejua committed:
:recycle: refactor: 메서드명 변경 PostImages의 isNotEmpty() -> IsEmpty() (#20)

olivejua committed:
:recycle: refactor: S3UploadService, S3Uploader의 메서드명 변경 (#20)

참고링크:
https://github.com/team-simplelog/simplelog/pull/20#discussion_r755815768

viiviii reviewed (approved):
리뷰 끝났는데 계속 리뷰 대기중이라고 뜨길래 확인해보니까 제가 `Comment`로 Review 제출해서였더라구요!
제가 원하던 건 `Approve`라 이제서야 변경했습니다🥲

각각 아래와 같이 쓰이나봐요. 
- Comment: Approve 없이 코멘트만 추가
- Approve: 와 좋아요 !우리 merge 해요! 느낌
- Request change: 꼭 이 부분은 바꿔주세요🥳

여러분은 다음 리뷰어 때  덜 헤매길 바래용😂

![image](https://user-images.githubusercontent.com/75404713/143367503-7c4c891c-abec-4200-ab2b-c098e2a9133e.png)

olivejua committed:
:recycle: refactor: S3UploadService의 Remove(User), Remove(Post) 리팩토링 (#20)

내용:
결합도 낮추는 목적
- User: getProfile() -> getProfileImageUrl() 메서드 변경 / hasProfileImage() 메서드 추가
- Post:  getImage() -> getImageUrls() 메서드 변경 / hasImages() 메서드 추가
- S3UploadService: User, Post의 변경된 메서드 호출하는 곳에 반영
가독성 목적
- 주석 추가: Post, User의 메서드들을 각 목적별로 분리 후 주석 추가
- S3UploadService: 지역변수 key를 path 로 변경

참고링크:
https://github.com/team-simplelog/simplelog/pull/20#discussion_r755816679

olivejua committed:
:recycle: refactor: User, Post의 임베디드 객체 리팩토링 (#20)

내용:
- User.profile: immutable 객체로 변환, User 생성자에서 파라미터 타입 변경 ProfileMessage -> Profile
- Post.tags, Post.images: Post 생성자에서 제외
- PostTags, PostImages: 참조변수 List를 final로 변환

참고링크:
- https://github.com/team-simplelog/simplelog/pull/20#discussion_r755847257
- https://github.com/team-simplelog/simplelog/pull/20#discussion_r755859057

olivejua committed:
:recycle: refactor: Bean AmazonS3Client -> AmazonS3 변경 (#20)

결합도 낮추는 목적
- (S3Uploader -> 구현클래스(AmazonS3Client)  =>  S3Uploader -> 인터페이스(AmazonS3))

참고링크:
- https://github.com/team-simplelog/simplelog/pull/20#discussion_r755774507

olivejua committed:
:recycle: refactor: S3UploadService 책임 변경, ImageUtils 클래스명 변경 및 메서드명 변경 (#20)

내용:
- S3UploadService 책임 변경: User의 프로필 이미지와 Post 첨부이미지 업로드 및 삭제 -> 이미지 업로드 및 삭제
- ImageUtils 클래스명 변경: ImageUtils -> DomainImagePaths
- ImageUtils 리팩토링: 각 도메인별 경로 생성 메서드 추가
- Post, User Path: url -> path 문자열 조작 로직을 객체 내부로 이동
- UploadServiceTest, DomainFilePathsTest: 위의 변경된 내용 반영

olivejua commented:
코드리뷰 첫경험이라... 리뷰내용을 실수로 모두 merge한 관계로 comment에 변경내용과 이유에 대해서 작성하겠습니다!

`S3UploadService.java`  이전 책임: `Post`와 `User`의 이미지 업로드 및 삭제한다.
- 구현했던 이유: 
구현당시 제 생각은 호출하는 `Service`에서 `File`의 경로형식이라던지 업로드 후 처리하는 방법을 service에서 다 해주면 책임이 더 커지지 않을까? 라는 생각에 File 업로드 및 삭제에 관련된 처리들을 해주었습니다.
 
리뷰 후 변경 이유: 
- uploadService라고 하면 upload 기능만 할 것 같은데 엔티티의 속성까지 조작하는 것이 예상에서 어긋난다는 점. (discussions: https://github.com/team-simplelog/simplelog/pull/20#discussion_r755913027 ), 
- Post와 User 이외에 다른 이미지 업로드 시 메서드가 늘어날 수 있다는 점

변경 후
`S3UploadService.java`  책임: 이미지 업로드 및 삭제한다.
- 내용: 이전 구현이유였던 경로 조작로직은 `uploadService.java` -> `DomainImagePaths.java`에 옮겼고, 각 도메인에 맞게 경로 조작하는 로직은 각 도메인의 내부 객체에서 담당하도록 변경했습니다!

ps. `DomainImagePaths` 가 이름이 딱 맘에 들지는 않지만 현재 생각나는 이름은 이것밖에 없었습니다..


`S3UploadService.java`  가 주로 변경이 많이 되어서 추가 설명이 필요할 것 같아 적었어여,,,
변경하니 비로소 제 옷을 입은듯... 의도와 맞아떨어지는 느낌이어서 아주 상쾌했습니다.....
병합이 이미 되어버렸지만 커밋내용 수정해야할 사항이 있다면 말해주세욥..

viiviii commented:
와... 개떡같이 말해도 찰떡같이 해내는 YOU✨,,,,
우리의 생애 첫 리뷰 경험 저도 넘나 유익 했습니다 👏
![image](https://user-images.githubusercontent.com/75404713/143411794-89791cf6-e0f9-4f01-a2f8-681363b50cfa.png)
